src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportProcessingJob.cs

```diff
@@ -74,11 +78,29 @@ public Task<string> ExecuteAsync(JobInfo jobInfo, IProgress<string> progress, Ca
                 TaskContinuationOptions.None,
                 TaskScheduler.Current);
 
-            Task<ExportJobOutcome> UpdateExportJob(ExportJobRecord exportJobRecord, WeakETag weakETag, CancellationToken innerCancellationToken)
+            async Task<ExportJobOutcome> UpdateExportJob(ExportJobRecord exportJobRecord, WeakETag weakETag, CancellationToken innerCancellationToken)
             {
                 record = exportJobRecord;
                 progress.Report(JsonConvert.SerializeObject(exportJobRecord));
-                return Task.FromResult(new ExportJobOutcome(exportJobRecord, weakETag));
+
+                if (record.Status != OperationStatus.Completed && record.Status != OperationStatus.Failed)
+                {
+                    // Update Export Job is called in three scenarios:
+                    // 1. A page of search results is processed and there is a continuation token.
+                    // 2. A filter has finished being used by a search
+                    // 3. The export job has reached a terminal state
+                    // This section is only reached in cases 1 & 2.
+                    // For jobs running in parallel such that each job only processes one page of results this section will never be reached as there are no filters and no continuation tokens.
+                    // For single threaded jobs this section will complete a job after every page of results is processed and make a new job for the next page.
+                    // This will allow for saving intermediate states of the job.
+
+                    var definition = JsonConvert.DeserializeObject<ExportJobRecord>(jobInfo.Definition);
+                    definition.Progress = exportJobRecord.Progress;
+                    await _queueClient.EnqueueAsync((byte)QueueType.Export, new[] { JsonConvert.SerializeObject(definition) }, jobInfo.GroupId, false, false, innerCancellationToken);
+                    throw new JobSegmentCompletedException();
```

Comment:
This is a little strange to me. I don't think I've seen success via exception before :)

Reply:
It isn't the best pattern I will admit, but it makes this still work with the old Cosmos system. Once your PR is finished so Cosmos uses the new JobHosting I can refactor ExportJobTask to handle this better.